### PR TITLE
Enum fetch more optimizations

### PR DESCRIPTION
**Pull request by eksperimental**
This PR does two optimizations,

First commit adds a clause to check for lists, when dealing with negative indexes
Second commit, revert's a change by @lexmag,
I think lists should be prioritized in Enum over ranges. In addition to this, I'm doing it based on the idea that checking for a range in a guard is more expensive than checking for a list, so it's better to leave it for as late as possible (just before the fallback guard that would deal with any other enumerable).

Benchmarks accompany this PR, between each commit.
https://gist.github.com/eksperimental/3f9c58a5b27fa9ff05abcd1df4678b1a

for the final results before this PR and after.
https://gist.github.com/eksperimental/3f9c58a5b27fa9ff05abcd1df4678b1a#file-enum_fetch_benchmarking_report_new-txt-L1172-L1360

the gain is considerable for out-of-bound and negative indexes.

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -721,38 +721,49 @@ defmodule Enum do
...
+    fetch_range(first, last, index)
+  end
+
   def fetch(enumerable, index) when is_integer(index) do
     count_result =
       case Enumerable.count(enumerable) do
-        {:ok, count} when (count - 1 - index) < 0 ->
+        {:error, module} ->
+          {:ok, module}
+        {:ok, count} when count <= index ->
```
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: this is just a simplification of the previous code.

**Comment by josevalim** [comment]
Assigned to @lexmag.

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -721,38 +721,49 @@ defmodule Enum do
...
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:ok, count} when (count + index) < 0 ->
-        :error
-
-      {:ok, count} ->
-        fetch_enumerable(enumerable, count + index, Enumerable)
-
-      {:error, _} ->
+    case enumerable do
```
Comment (ericmj) [inline lib/elixir/lib/enum.ex]: I think it's clearer to use function clauses rather than a `case`. The rest of the module uses function clauses to dispatch to different optimizations.

**Comment by lexmag** [comment]
> I think lists should be prioritized in Enum over ranges. In addition to this, I'm doing it based on the idea that checking for a range in a guard is more expensive than checking for a list, so it's better to leave it for as late as possible (just before the fallback guard that would deal with any other enumerable).

The performance difference is so negligible (if any) that we should not consider it over readability, compiler can freely regroup matching clauses by type and do other rewrites that have no affect on the program flow. The cost for checking a range in guard is simply the cost for checking a map (Range struct).

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -747,12 +755,12 @@ defmodule Enum do
   def fetch(enumerable, index) when is_integer(index) do
     count_result =
       case Enumerable.count(enumerable) do
-        {:ok, count} when (count - 1 - index) < 0 ->
+        {:error, module} ->
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Let's to not change the order. :bowtie:

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -764,15 +775,15 @@ defmodule Enum do
   end
 
   defp fetch_enumerable(enumerable, index, module) do
-    reduce_result =
+    reduce_res =
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Unnecessary change.

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Agree with @ericmj, we can move all list handling in one place:

```elixir
  def fetch(enumerable, index) when is_list(enumerable) and is_integer(index) do
    if index < 0 do
      enumerable
      |> :lists.reverse
      |> fetch_list((-index) - 1)
    else
      fetch_list(enumerable, index)
    end
  end
```

File: lib/elixir/lib/enum.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: it was done after these changes to the code I previously submitted.
- https://github.com/elixir-lang/elixir/commit/4b3b4370d8b821d92a368d5f58c1247044092473#diff-6881431a92cd4e3ea0de82bf2338f8eaL754
- https://github.com/elixir-lang/elixir/commit/4b3b4370d8b821d92a368d5f58c1247044092473#diff-6881431a92cd4e3ea0de82bf2338f8eaL773

File: lib/elixir/lib/enum.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: oops!

File: lib/elixir/lib/enum.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: Sorry It got messed up, probably because I based these changes on code previous to @lexmag correction.
change reverted, all good then

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: @ericmj, that is definitely the way to go for this case.
you suggestion simplifies the code A LOT and solves the issues related to the performace addressed by the PR.
thank you!!!

**Comment by eksperimental** [comment]
@lexmag may I ask what is considered negligible and above what limit is not.
I think the performance gains obtained  by this change are not negligible.
you can see them in the first comparison
https://gist.github.com/eksperimental/3f9c58a5b27fa9ff05abcd1df4678b1a#file-enum_fetch_benchmarking_report_new-txt-L977-L1165
in some cases with lists, gain goes from 5-9%.
all cases in map gain upt 7-16% in the best case.
allmost in all structs the gain is of 15%.

anyway this has been solved by reorganizing the code, but I would like to know for better performance related PRs

thanks for your suggestions, always welcome and appreciated

**Comment by whatyouhide** [comment]
I would not focus too much on gains for maps/structs as, as stated somewhere else in the past, `Enum.fetch/2` is not really meant for maps/structs as the keys are not ordered (and in my code, I have never used `fetch/2` on maps or structs).

**Comment by eksperimental** [comment]
dang it! i keep forgetting about those @whatyouhide

**Comment by lexmag** [comment]
@eksperimental it was about "I think lists should be prioritized in Enum over ranges.", the difference between lists-matching-first and ranges-matching-first is negligible.

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -721,45 +721,33 @@ defmodule Enum do
...
-      case Enumerable.count(enumerable) do
-        {:ok, count} when (count - 1 - index) < 0 ->
-          :error
-        {:ok, _} ->
-          {:ok, Enumerable}
-        {:error, module} ->
-          {:ok, module}
-      end
+    case Enumerable.count(enumerable) do
+      {:error, module} when index < 0 ->
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Each of these patterns have special handling for negative `index`, I think we should have a separate function clause for it similarly as we had before.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: more code, but easier to read :+1:

**Comment by lexmag** [comment]
Thank you @eksperimental. 💛
